rabbitmq/resource_queue.go

```diff
@@ -132,12 +132,33 @@ func ReadQueue(d *schema.ResourceData, meta interface{}) error {
 	e := make(map[string]interface{})
 	e["durable"] = queueSettings.Durable
 	e["auto_delete"] = queueSettings.AutoDelete
-	e["arguments"] = queueSettings.Arguments
-	queue[0] = e
 
-	d.Set("settings", queue)
+	// NOTE the addition of `arguments_json` never actually worked
+	// https://github.com/terraform-providers/terraform-provider-rabbitmq/pull/6
+	// d.Set("settings", queue) failed silently as a result of not checking the error.
+	// Additionally there would be a permadiff between the configuration and schema
+	// as the queue previously always set `arguments` on read. Issue was filed
+	// https://github.com/terraform-providers/terraform-provider-rabbitmq/issues/16
+	// In order to suppress the permadiff and error at the same time, if the
+	// data originated from `arguments_json` store the data in that key as a
+	// json string. We can feel confident in the data's origin because the schema
+	// was setup to conflict if both `arguments` and `arguments_json` were present
+	// This fix will not account for the case where `arguments` was used to supply
+	// string arguments to a queue, then the queue had arguments added outside of terraform,
+	// with values others than string, then d.Set("settings", queue) would fail.
+	if _, ok := d.GetOk("settings.0.arguments_json"); ok {
+		bytes, err := json.Marshal(queueSettings.Arguments)
+		if err != nil {
+			return err
+		}
+		e["arguments_json"] = string(bytes)
+	} else {
+		e["arguments"] = queueSettings.Arguments
```

Comment:
Is it possible that drift in the remote system could cause `queueSettings.Arguments` to have values in it that aren't strings? If that were true, I think we'd leave this block with an invalid value for `arguments` that `d.Set("settings", queue)` would reject, leaving us with the unchanged prior value in the state.

Perhaps instead we could check whether everything in `queueSettings.Arguments` is a string, and if not then force the value to be written into `arguments_json` regardless of what was already in the state. A non-string value can never be represented in `arguments` anyway, so there should be no situation where a config using `arguments` should produce something that can't be written back to `arguments` in the absence of drift, but if drift _does_ introduce something unrepresentable then writing it into `arguments_json` at least allows us to show a subsequent plan diff with all of the information, rather than failing to detect the drift at all. In the 0.12 plan output it would look something like this:

```
    - arguments_json = jsonencode({
        originally_set = "foo"
        drift_value    = ["foo", "bar"]
      })
    + arguments = {
        originally_set = "foo"
      }
```

Assuming a configuration containing this, with `drift_value` being something introduced outside of Terraform since the object was created:

```hcl
  arguments = {
    originally_set = "foo"
  }
```

It's unfortunate that we'd show it as removing `arguments_json` and adding `arguments` rather than just showing an in-place update for one of them, but I think this is better than not detecting any change at all in that case, which I think is what would happen with this proposed logic.

Reply:
Yes i mention that in my novel length comment, drift to 'arguments' introducing non strings will break. I like the suggestion you have made of forcing such cases into arguments_json. We will still have drift but at least the provider is fully functional again.

Reply:
Oh, hmm... I was thinking of more like a combination of what you had it doing before and what you have it doing now: if `arguments_json` is set in the state then use it unconditionally, but if `arguments` is set in the state then use it only if the remote value contains only strings. Otherwise I think a use of `arguments_json` that only sets strings would cause a permadiff, right?

```
  arguments_json = jsonencode({"foo" = "bar"})
```

Reply:
🤦‍♂ Yes you are right, exhaustive acceptance tests would catch these edgecases, a lesson learned to write the tests first...